stored_requests/config/config.go

```diff
@@ -22,48 +22,152 @@ import (
+		if conn != dbConnection.conn {
+			glog.Fatal("Multiple database connection settings found in Stored Requests config, only a single database connection is currently supported.")
+		}
+	}
+
+	eventProducers := newEventProducers(cfg, client, db, router)
+	fetcher = newFetcher(cfg, client, db)
+
+	var shutdown1 func()
+
+	if cfg.InMemoryCache.Type != "" {
+		cache := newCache(cfg)
+		fetcher = stored_requests.WithCache(fetcher, cache)
+		shutdown1 = addListeners(cache, eventProducers)
+	}
+
 	shutdown = func() {
-		shutdown1()
-		shutdown2()
-		if db != nil {
+		if shutdown1 != nil {
+			shutdown1()
+		}
+		if dbConnection.db != nil {
+			db := dbConnection.db
+			dbConnection.db = nil
+			dbConnection.conn = ""
 			if err := db.Close(); err != nil {
 				glog.Errorf("Error closing DB connection: %v", err)
 			}
 		}
 	}
+
+	return
+}
+
+// NewStoredRequests returns five things:
+//
+// 1. A DB connection, if one was created. This may be nil.
+// 2. A function which should be called on shutdown for graceful cleanups.
+// 3. A Fetcher which can be used to get Stored Requests for /openrtb2/auction
+// 4. A Fetcher which can be used to get Stored Requests for /openrtb2/amp
+// 5. A Fetcher which can be used to get Category Mapping data
+//
+// If any errors occur, the program will exit with an error message.
+// It probably means you have a bad config or networking issue.
+//
+// As a side-effect, it will add some endpoints to the router if the config calls for it.
+// In the future we should look for ways to simplify this so that it's not doing two things.
+func NewStoredRequests(cfg *config.Configuration, client *http.Client, router *httprouter.Router) (db *sql.DB, shutdown func(), fetcher stored_requests.Fetcher, ampFetcher stored_requests.Fetcher, categoriesFetcher stored_requests.CategoryFetcher) {
+	// Build individual slim options from combined config struct
+	slimAuction, slimAmp := resolvedStoredRequestsConfig(cfg)
+
+	// TODO: Switch this to be set in config defaults
+	//if cfg.CategoryMapping.CacheEvents.Enabled && cfg.CategoryMapping.CacheEvents.Endpoint == "" {
+	//	cfg.CategoryMapping.CacheEvents.Endpoint = "/storedrequest/categorymapping"
+	//}
+
+	fetcher1, db, shutdown1 := CreateStoredRequests(&slimAuction, client, router)
+	fetcher2, _, shutdown2 := CreateStoredRequests(&slimAmp, client, router)
+	fetcher3, catdb, shutdown3 := CreateStoredRequests(&cfg.CategoryMapping, client, router)
+
+	// Return the database still should it be defined in CategoryMapping and not StoredRequests in the config object.
+	if db == nil && catdb != nil {
+		db = catdb
+	}
+
+	fetcher = fetcher1.(stored_requests.Fetcher)
+	ampFetcher = fetcher2.(stored_requests.Fetcher)
+	categoriesFetcher = fetcher3.(stored_requests.CategoryFetcher)
+
+	shutdown = func() {
+		shutdown1()
+		shutdown2()
+		shutdown3()
+	}
+
+	return
+}
+
+func resolvedStoredRequestsConfig(cfg *config.Configuration) (auc, amp config.StoredRequestsSlim) {
+	sr := &cfg.StoredRequests
```

Comment:
If we only use `sr`, and not the original `cfg`, it may be best to pass just that.